File: agents/adapters/llm/src/llm_adapter/handler.py
# SPDX-License-Identifier: Apache-2.0
"""ChatCompletionHandler — validates input, streams provider tokens, emits TaskEvents."""

from __future__ import annotations

import asyncio
import json
from collections.abc import AsyncGenerator
from datetime import UTC, datetime

from google.protobuf.timestamp_pb2 import Timestamp  # type: ignore[import-untyped]
from pydantic import BaseModel, Field, ValidationError
from zynax.v1 import agent_pb2  # type: ignore[import-untyped]

from llm_adapter.config import ProviderConfig
from llm_adapter.providers import ProviderFunc, get_provider

_MAX_ERR = 512


class _ChatInput(BaseModel):
    """Validated chat completion input payload."""

    prompt: str = Field(..., min_length=1)
    system: str | None = None
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    max_tokens: int | None = Field(default=None, ge=1)


def _ts() -> Timestamp:
    """Return the current UTC time as a protobuf Timestamp."""
    ts = Timestamp()
    ts.FromDatetime(datetime.now(UTC))
    return ts


def _progress(task_id: str, chunk: bytes) -> agent_pb2.TaskEvent:
    """Build a PROGRESS TaskEvent."""
    return agent_pb2.TaskEvent(
        task_id=task_id,
        event_type=agent_pb2.TASK_EVENT_TYPE_PROGRESS,
        payload=chunk,
        timestamp=_ts(),
    )


def _completed(task_id: str, full_text: str) -> agent_pb2.TaskEvent:
    """Build a COMPLETED TaskEvent with the full response payload."""
    return agent_pb2.TaskEvent(
        task_id=task_id,
        event_type=agent_pb2.TASK_EVENT_TYPE_COMPLETED,
        payload=json.dumps({"response": full_text}).encode(),
        timestamp=_ts(),
    )


def _failed(task_id: str, code: str, message: str) -> agent_pb2.TaskEvent:
    """Build a FAILED TaskEvent with a sanitised error message."""
    return agent_pb2.TaskEvent(
        task_id=task_id,
        event_type=agent_pb2.TASK_EVENT_TYPE_FAILED,
        error=agent_pb2.CapabilityError(code=code, message=message[:_MAX_ERR]),
        timestamp=_ts(),
    )


def _parse_input(payload: bytes) -> _ChatInput:
    """Decode and validate the input payload JSON.

    Args:
        payload: Raw bytes from ``ExecuteCapabilityRequest.input_payload``.

    Returns:
        Validated ``_ChatInput`` instance.

    Raises:
        ValueError: On JSON decode failure or schema validation error.
    """
    try:
        data = json.loads(payload)
    except (json.JSONDecodeError, UnicodeDecodeError) as exc:
        raise ValueError(f"invalid JSON in input_payload: {exc}") from exc
    try:
        return _ChatInput.model_validate(data)
    except ValidationError as exc:
        raise ValueError(str(exc)) from exc


class ChatCompletionHandler:
    """Handles chat_completion capability execution.

    Validates the input payload, invokes the configured provider, and
    streams PROGRESS events followed by exactly one terminal event.
    """

    def __init__(self, config: ProviderConfig) -> None:
        """Initialise with provider configuration.

        Args:
            config: Validated provider configuration; provider callable is
                built once at construction time.
        """
        self._max_tokens: int = config.max_tokens
        self._provider: ProviderFunc = get_provider(config)

    async def stream(
        self,
        task_id: str,
        input_payload: bytes,
        timeout_seconds: int,
    ) -> AsyncGenerator[agent_pb2.TaskEvent, None]:
        """Execute the capability and stream TaskEvents to the caller.

        Args:
            task_id: Unique task identifier; echoed on every event.
            input_payload: JSON-encoded ``_ChatInput`` bytes.
            timeout_seconds: Total wall-clock budget for the provider call.

        Yields:
            PROGRESS events (one per token chunk), then exactly one terminal
            COMPLETED or FAILED event.
        """
        try:
            inp = _parse_input(input_payload)
        except ValueError as exc:
            yield _failed(task_id, "INVALID_INPUT", str(exc))
            return

        max_tokens = inp.max_tokens or self._max_tokens
        chunks: list[bytes] = []
        emitted = False

        try:
            async with asyncio.timeout(float(timeout_seconds)):
                async for chunk in self._provider(
                    inp.prompt, inp.system, inp.temperature, max_tokens
                ):
                    ev = _progress(task_id, chunk)
                    yield ev
                    chunks.append(chunk)
                    emitted = True
        except TimeoutError:
            if not emitted:
                yield _progress(task_id, b"")
            yield _failed(task_id, "TIMEOUT", "request exceeded timeout")
            return
        except RuntimeError as exc:
            if not emitted:
                yield _progress(task_id, b"")
            yield _failed(task_id, "UPSTREAM_ERROR", str(exc))
            return

        if not emitted:
            yield _progress(task_id, b"")
        yield _completed(task_id, b"".join(chunks).decode(errors="replace"))

File: agents/adapters/llm/internal/domain/handler.go
// SPDX-License-Identifier: Apache-2.0

package domain

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/santhosh-tekuri/jsonschema/v5"
	"github.com/zynax-io/zynax/agents/adapters/llm/internal/provider"
	zynaxv1 "github.com/zynax-io/zynax/protos/generated/go/zynax/v1"
)

// maxErrMsgLen bounds any CapabilityError.message so a verbose upstream or
// validation error can never bloat a terminal event (canvas M7.P safeguards).
const maxErrMsgLen = 512

// Well-known capability error codes (canvas E / parity oracle).
const (
	codeInvalidInput = "INVALID_INPUT"
	codeTimeout      = "TIMEOUT"
	codeUpstream     = "UPSTREAM_ERROR"
)

// EventSink is the minimal surface of AgentService_ExecuteCapabilityServer the
// handler needs. Defining it here keeps the handler testable without a real gRPC
// server and decouples domain logic from the transport.
type EventSink interface {
	Send(*zynaxv1.TaskEvent) error
	Context() context.Context
}

// chatInput is the validated chat_completion input payload. Only prompt is
// required; provider, model, and credentials are never read from the payload
// (canvas safeguards) — they come from static config.
type chatInput struct {
	Prompt string `json:"prompt"`
}

// ChatCompletionHandler validates input_payload against the declared JSON Schema,
// invokes the configured Provider, streams per-token PROGRESS events, and emits
// exactly one terminal event. It is stateless and safe for concurrent use.
type ChatCompletionHandler struct {
	provider provider.Provider
	schema   *jsonschema.Schema // compiled input schema; nil when none declared
}

// newChatCompletionHandler binds a Provider and compiles the declared input
// schema once at startup. An empty schema string disables structural validation
// (prompt presence is still enforced at request time).
func newChatCompletionHandler(p provider.Provider, inputSchemaJSON string) (*ChatCompletionHandler, error) {
	h := &ChatCompletionHandler{provider: p}
	if strings.TrimSpace(inputSchemaJSON) == "" {
		return h, nil
	}
	compiler := jsonschema.NewCompiler()
	if err := compiler.AddResource("input.json", strings.NewReader(inputSchemaJSON)); err != nil {
		return nil, fmt.Errorf("domain: add input schema: %w", err)
	}
	sch, err := compiler.Compile("input.json")
	if err != nil {
		return nil, fmt.Errorf("domain: compile input schema: %w", err)
	}
	h.schema = sch
	return h, nil
}

// Execute runs the capability and streams TaskEvents on sink: PROGRESS per token
// (at least one), then exactly one terminal COMPLETED or FAILED event. task_id
// and a timestamp are set on every event. timeoutSeconds bounds the provider
// call via a derived context deadline. The returned error is only a transport
// (Send) failure; capability failures are delivered as a FAILED event.
func (h *ChatCompletionHandler) Execute(taskID string, payload []byte, timeoutSeconds int32, snk EventSink) error {
	s := streamSink{EventSink: snk}
	prompt, err := h.parseInput(payload)
	if err != nil {
		return s.send(failedEvent(taskID, codeInvalidInput, err.Error()))
	}

	streamCtx := snk.Context()
	if timeoutSeconds > 0 {
		var cancel context.CancelFunc
		streamCtx, cancel = context.WithTimeout(streamCtx, time.Duration(timeoutSeconds)*time.Second)
		defer cancel()
	}

	chunks, err := h.provider.Stream(streamCtx, prompt)
	if err != nil {
		return s.terminalForError(streamCtx, taskID, err, false)
	}
	return h.relay(streamCtx, s, taskID, chunks)
}

// relay pumps provider chunks to PROGRESS events and emits the terminal event.
// It guarantees at least one PROGRESS even for an empty stream so the contract
// "at least one PROGRESS before terminal" always holds.
func (h *ChatCompletionHandler) relay(streamCtx context.Context, s streamSink, taskID string, chunks <-chan provider.Chunk) error {
	var b strings.Builder
	emitted := false
	for c := range chunks {
		if c.Err != nil {
			return s.terminalForError(streamCtx, taskID, c.Err, emitted)
		}
		if err := s.send(progressEvent(taskID, []byte(c.Text))); err != nil {
			return err
		}
		b.WriteString(c.Text)
		emitted = true
	}
	if streamCtx.Err() != nil {
		return s.terminalTimeout(taskID, emitted)
	}
	if err := s.ensureProgress(taskID, emitted); err != nil {
		return err
	}
	return s.send(completedEvent(taskID, b.String()))
}

// parseInput decodes and validates payload against the declared schema, then
// extracts the prompt. Any failure maps to INVALID_INPUT.
func (h *ChatCompletionHandler) parseInput(payload []byte) (string, error) {
	var v interface{}
	if err := json.Unmarshal(payload, &v); err != nil {
		return "", fmt.Errorf("invalid JSON in input_payload: %w", err)
	}
	if h.schema != nil {
		if err := h.schema.Validate(v); err != nil {
			return "", fmt.Errorf("input_payload schema validation: %w", err)
		}
	}
	var in chatInput
	if err := json.Unmarshal(payload, &in); err != nil {
		return "", fmt.Errorf("invalid input_payload: %w", err)
	}
	if in.Prompt == "" {
		return "", fmt.Errorf("input_payload field \"prompt\" is required")
	}
	return in.Prompt, nil
}
